src/casp/services/api/services/cell_operations_service.py

```diff
@@ -58,6 +62,51 @@ def authorize_model_for_user(self, user: models.User, model_name: str) -> models
             )
         return model
 
+    def __get_cell_count_from_anndata(self, file: t.BinaryIO) -> int:
+        """
+        Get the number of cells in the anndata file.
+
+        :param file: Anndata file to get the number of cells from.
+
+        :return: Number of cells in the anndata file.
+        """
+        adata = anndata.read_h5ad(file)
+        cell_count = len(adata)
+        # Reset the file pointer to the beginning of the file because we're gonna need to read it
+        # again later.  I don't know if this is the cleanest solution to this problem
+        file.seek(0)
+        return cell_count
+
+    def __verify_quota_and_log_activity(
+        self, user: models.User, file: t.BinaryIO, model_name: str, method: str
+    ) -> None:
+        """
+        Verify that the number of cells in the anndata file does not exceed the user's remaining
+        quota and log the user activity in the database in an atomic transaction.
+
+        :param user: User object to check quota for.
+        :param file: Anndata file to check the number of cells in.
+        :param model_name: Model name to use for annotation.
+        :param method: Method name to log in the database.
+
+        :raises exceptions.QuotaExceededException: If the number of cells in the anndata file exceeds the user's quota.
+        """
+        cell_count = self.__get_cell_count_from_anndata(file=file)
+        with user_activity_lock:
```

Comment:
Ah, good to know that's how to do locks!  I'll just mention that this is only per instance so if we're running several replicas, this won't actually lock

Reply:
Oh I didn't think of that.  That does complicate things.  I'm not sure how to handle it in that case 🤔